### ISSUE TITLE
bedmap --chrom with piped refernce file

### ISSUE THREAD

**Issue by bgulko** [bedops/bedops#136]
When reading data from a ref-file that is a pipe, the --chrom option causes chromosomes to be ignored. If the argument to --chrom is not the first chromosome, then no output is generated.

This works as expected:
 bedmap  --chrom chr2 --skip-unmapped --echo --echo-map-id-uniq r.bed m.bed
chr2    28117   28118   M       0.999964                        |0N0NN0130
chr2    28127   28128   M       0.999914                        |0N0NN0130

But adding the --chrom filter results in no loci output:
 bedmap  --chrom chr2 --skip-unmapped --echo --echo-map-id-uniq <( cat r.bed ) m.bed

While leaving out the --chrom option, but keeping the pipe, works fine:
 bedmap  --skip-unmapped --echo --echo-map-id-uniq <(cat r.bed )  m.bed
chr1    4091    4092    M       0.977821                        |0N0NN0130
chr2    28117   28118   M       0.999964                        |0N0NN0130
chr2    28127   28128   M       0.999914                        |0N0NN0130

m.bed
chr1    0       10000   0N0NN0130
chr11   67471   67546   0N1NN0130
chr2    0       10200   0N0NN0130
chr2    19200   41610   0N0NN0130
chr2    41610   41627   0N0NN0131
r.bed
chr1    4091    4092    M       0.977821
chr11   190783  190784  M       0.999914
chr2    28117   28118   M       0.999964
chr2    28127   28128   M       0.999914

**Comment by bgulko** [comment]
This is bedmap 2.4.11 for linux x64.

**Comment by alexpreynolds** [comment]
Shane will likely be able to confirm or correct me, but I think the `--chrom` functionality requires a regular file.

This is because finding boundaries between chromosomes requires the ability to do random seeks (`fseek()` calls) to search backwards and forwards through a file, and that isn't possible with a data stream coming from standard input. A pipe only lets you look at whatever chunk of data has come in.

If I'm correct, then we'll make some documentation changes and perhaps a patch to exit `bedmap` with a warning if `stdin` is specified with this option.

**Comment by alexpreynolds** [comment]
Note that you can do the following, however:

```
$ bedextract chr2 r.bed | bedmap --skip-unmapped --echo --echo-map-id-uniq - m.bed
```

I think `bedmap` may need to stream through `m.bed` up until the last `chr2` record in `r.bed`, so it won't be as fast as applying `--chrom` to `bedmap` directly, though it should still be more efficient than streaming through all of `r.bed`.

Note that `bedextract` does not allow `stdin` with a chromosome name, for the same reason as described above.

**Comment by sjneph** [comment]
bedmap requires that you use a dash (-) to tell it which file is coming from stdin.  This works in both tcsh and bash.
bedmap --skip-unmapped --chrom chr2 --echo --echo-map-id-uniq - m.sort.bed < r.sort.bed

That probably solves your problem.  I am curious though, what shell environment are you using?

As far as using the --chrom operator with bedmap, there is definitely a possible penalty if you're getting the input from stdin.  It will work, but the data will be streamed through until the first record after the chromosome of interest.  If you give it a file input, as Alex wrote about, it will jump to chr2 as efficiently as bedextract does.

Let us know if the above syntax doesn't fix the issue.

**Comment by bgulko** [comment]
Thanks for the quick response! The workaround does indeed
produce the correct results for this debugging example.

An explicit error message, rater than a quiet failure
would ameliorate most of the impact of this problem,
which cost me a day to track down in a complex pipeline.

Personally, my pipeline often requires two piped sources
of input to prevent the expensive writing of intermediate
files to disk.

It is certainly true that piped outputs reduce the
efficacy of chromosomal indexing. However, I have found
that costs of ignoring large segments of an input stream
be offset by the decompression penalty incurred by in
chromosome-indexed files like starch (likely due
to aggressive file caching of streamed data by the OS).

I recently found that, for 50GB bed file:
	Streaming the whole file (22 chromosomes) takes less than a minute
	Using starch, Skipping to chromosome 1 and decompressing
		just that chromosome took
		> 3 min with bzip2 compression
		~ 1 min with gzip  compression

Surprisingly, it seems that sometimes it is faster to stream
the uncompressed file from disk, than to decompress large
portions of it, even if this stream has to be processed for
each parallel job thread (per chromosome).

--Brad

**Comment by bgulko** [comment]
Sorry, looks like I read these out of order.

I think the combination of an error for a <() and a
warning for redirection from stdin would be perfect.

In any case, thank you for this tremendous tool, really,
I've used it nearly every day for several years.

--Brad

**Comment by bgulko** [comment]
Hi Shane

I'm using bash with sub-process redirection.

The basic arrows (  < and >  ) are generally used with
standard streams (stdin, stdout, stderr). This is great when
you only want to redirect one or two streams. When you
want to combine multiple streams from sub-processes into
one command (say "paste"), without writing the results to disk
this bash syntax does it....

paste <( command1 ) <( command2) <( command3 ).....

This runs each command, and generates a pipe-end for the stdout
for the command and provides that pipe-end to the "paste"
command as a standard file name (usually something like /dev/fd/XX).
When "paste" reads from the "file" it gets the output stream from the
command.

I use this a lot, especially when some minor surgery, like locus tagging,
has to be added to a large database/bed file, without rewriting the whole
file. It allows a static file to be replaced with some simple functional
modification to that file, without modifying a potentially complex,
underlying pipeline.

--Brad

**Comment by alexpreynolds** [comment]
One minute is a pretty fast stream time through a 50 GB file. I'd suggest double-checking that those times aren't from cached data. 

If you're running Linux, you can run the following as root to reset the cache:

```
$ echo 3 > /proc/sys/vm/drop_caches
```

Otherwise, if the file isn't cached, you should see an overall benefit from getting data from an indexed starch file, especially with parallelized work. If that isn't the case, I'd definitely like to know more about your setup and inputs.

**Comment by sjneph** [comment]
I'm sorry that you lost some time tracking things down.  I'd like to tell you that I haven't done the same more than once.  I am always happy to hear when someone out there is using the tools regularly.  I know Alex is too.

To be honest, I've never thought about the situation when someone pipes input into the program but doesn't actually use it.  In the case of these tools, you must use '-'.  Without that, stdin is ignored, and this is (part of) what happened here.  It is an interesting thought to pick up a possible use error like this though.  I'll have to look some things up, but it seems reasonable that we could look at stdin and ensure nothing is coming in when '-' is not used.

Thanks for the comments.

**Comment by sjneph** [comment]
You should be able to use your subprocesses and the same syntax that I wrote above - you are only missing the dash, I believe.  If that isn't true, let us know.

**Comment by bgulko** [comment]
I'm pretty sure large parts of the file ARE cached, even with SSD >800MB/s would
be suspicious. The point is that multiple processes sharing the same cached file
will share the benefits of streaming a no-op through the large cache.
The numbers were surprising to me as well, but it looks like uncompressed
data might be a net win (is there a "no compression" option for starch?).

The setup is a somewhat pathological case where we have multidimensional
values for individual genomic positions for around 1/2 of the human genome,
but the values are widely diffused so even moderately fine blocks will still
result in the entire genome (chromosome) having to be uncompressed at each
pass. I started assuming that decompression was inexpensive relative IO
times, but found that decompression seemed more of a bottleneck.

-- Brad

**Comment by bgulko** [comment]
Yes, I can use the syntax you propose, so long as I only need one pipe.
I often end up needing more than one input pipe, and that can presents
more of a problem.

As you mentioned, mine is an odd use case where (likely due to a large system file
cache) streaming uncompressed data seems faster than using compressed data with
chromosome-level indexing.

--Brad

**Comment by alexpreynolds** [comment]
As an alternative to `starch`, you might be interested in `bedextract`, which will let you extract a chromosome of data from an uncompressed BED file fairly quickly. 

For example, to extract `chrN` records from a `sort-bed`-sorted regular file called `elements.bed`, you could do the following:

```
$ bedextract chrN elements.bed > elements.chrN.bed
```

More details are available here:

http://bedops.readthedocs.org/en/latest/content/reference/set-operations/bedextract.html

**Comment by alexpreynolds** [comment]
If you want to parallelize work by chromosome, `bedextract` helps here, as well:

```
$ for chr in `bedextract --list-chr elements.bed`; \
    do \
        bedextract ${chr} elements.bed > /tmp/elements.${chr}.bed; \
        qsub /*options*/ do_work /tmp/elements.${chr}.bed ... ; \ 
    done
```

Or via GNU Parallel:

```
$ bedextract --list-chr elements.bed | parallel "bedextract {} elements.bed > /tmp/elements.{}.bed && do_work /tmp/elements.{}.bed ... "
```

**Comment by sjneph** [comment]
I use a lot of named pipes in my work when I'm trying to get around huge temp files.  You probably already realize this, but combining bedops -u with bedmap will solve the many input files problem for regular files and named pipes.

`bedops -u r1 r2 r3 ... | bedmap ... - m.bed`

I haven't really thought about the case of multiple regular pipes.  Using parallel as Alex suggests may be able to help there.

**Comment by bgulko** [comment]
Thanks Alex, I had started with the expectation that
I could use bedextract for this design, but ran into
some limitations. I'm segmenting a genomic database
with a separate entry at each genomic position, into a
number of classes (from 2 to 1000 or so) that together
span the entire genome. Each class contains many small
loci.

Approach 1 - run bedextract for each class.....
	bedextract only accepts a starched file for the target,
	not the querry, and the querry files can be quite large.
	Iteratively extracting dozens of large classes from a
	large database file proved prohibitive.

Approach 2 - Make a single streaming pass through the database (target)
	and tag each element with its class, then use awk to split the
	tagged database into one file per class. This has the advantage of only
	requiring one pass but requires that I propagate a tag from the query,
	as well as the data from the target, to the result file. Perfect for bedmap,
	but not really a strength of bedextract or bedops.

Approach2 seems a lot faster, but I have run into limitations because
both the target and the query are on the order of the size of the
entire human genome, with small individual loci (database is single
nucleotides and querry loci are 10's to 100 of bp), so both files
are large.

**Comment by sjneph** [comment]
If your files do not contain fully-nested elements (it sounds like you use --bp-ovr in this case), you should try out --faster with bedmap.  If you're unsure, adding --ec will tell you if that assumption is incorrect.  For the SNP file, you're safe so only evaluate the other file.  --ec will roughly double the run-time, but --faster can speed things up a great deal more than that under the right conditions.  Using --ec is up to you, but be mindful that --faster will produce incorrect results if the assumption is not met.

If you have access to a cluster, definitely use the --chrom operator.  If you're on a single machine with a few cores, gnu parallel can be helpful with --chrom.

You might take a look at starchcluster_sge.tcsh and starchcluster_gnuParallel.tcsh in the applications/bed/starch/src/ directory to see how we do these kinds of things.  The meat that uses a cluster looks like:
`@ cntr = 0
set files = ()
set jids = ()
foreach chrom (`bedextract --list-chr $input`)
qsub $sge_opts -N job.$cntr > /dev/stderr << __EXTRACTION__
  bedextract $chrom $input | starch - > $cntr
  set files = ($files $cntr)
  set jids = ($jids $job.cntr)
  @ cntr++
__EXTRACTION__`

and then unionizing the results

`qsub $sge_opts -hold_jid `echo $jids | tr ' ' ','` > /dev/stderr << __CATTED__
  starchcat $files > $output
__CATTED__`

In your set-up with starch file inputs and a starch file output, you would change the:
`bedextract --list-chr`
to
`unstarch --list-chr`

and
`bedextract $chrom $input | starch - > $cntr`
to
`bedmap --chrom $chrom --faster [other-operations...] $ref $map | starch - > $cntr`

The rest of the script just checks inputs and deals with creating a temp directory to work in and clean up.  starchcat is very efficient to use in this way to create your final archive.  The amount of space used during all of this is should be no more than the final output file size.

**Comment by sjneph** [comment]
Upon further review of the initial problem, there is something weird going on with --chrom.  I'm not very familiar with the bash subprocess substitution.  It looks similar to named pipes and perhaps that's how it works under the hood.
Looking at named pipes in tcsh, I get the same behavior.  If I do not use --chrom chr2, things work fine.  If I do, things break.  This is a confirmed bug.

**Comment by bgulko** [comment]
Thanks for the update, and sorry it turned out to be more problematic. I might
have a clue. Named pipes should work just like input file redirection. The major
difference between pipes and files, from an application's perspective, is that
a file has all the data an application can read, while a pipe is generating
date in real time. The result is that a read on a pipe may only return a fraction
of the data requested, even if there is more that will eventually be available.
A file, on the other hand, always returns the full amount of data requested unless
it has reached the end of a file.

A common problem for applications reading from pipes, is that they assume that
when the a read buffer returned is smaller than the amount of data requested, that
they have reached an "end of file". If this "end of file" is reached before the
end of a block (say a compression block) it is assumed that the stream is corrupt.

It often helps to put an i/o layer between the application's "read" command and the
system's pipe IO. This layer blocks when there is not enough data in the pipe to fulfill
the request, until either the pipe is closed (end of file) or the remainder of the
requested has arrived.

I don't know if this is the particular problem, but it might help,

Best,
--Brad

**Comment by sjneph** [comment]
Thanks.  Many of the utilities assume an input is either a file or it's stdin.  With named pipes, including bash's syntactic sugar for them, it fools the applications.  If it's stdin, then we stream through the data for the correct chromosome (inefficient).  Otherwise, we jump around using a binary search as a file input is assumed.  I just need to check whether it's a named pipe, which gets into icky OS-specific stuff that I normally try to avoid.  But, no problem to fix probably.

**Comment by sjneph** [comment]
Well, without --ec, things were straightforward.  With --ec, I'm having some troubles.  I'm unsure if I can easily fix this.  The subprocess syntax uses a named pipe, but there is something else going on too (probably exec/forking with subprocesses as the name implies).  I'm unable to successfully wrap it with a std::istream in C++ so far (what we use in --ec mode).  It knows it's a named pipe, but when I read from it, I get nothing.  Without --ec, we use a straight FILE* and that doesn't have any problems with this.

When testing with regular named pipes in tcsh, I had no problems with/out --ec.

**Comment by sjneph** [comment]
It's weirder than just talking about subprocess stuff.  The same thing (I thought) happens in tcsh with named pipes/subprocesses.
'''mfifo f1 f2'''
'''(cat m.bed > f1)&'''
'''(cat r.bed > f2)&'''
'''bedops -u --ec f1 f2 > f3'''
'''wait'''

this has no problem (and f1/f2 are wrapped in istream's, and coming from other processes).  I'm not sure.

**Comment by sjneph** [comment]
Okay, I mislead myself pretty good there.  The bash syntax is exactly what I figured.  I made a silly mistake.  Things seem to work as you want with process substitution now (this is something new to me).  Updates will be available when we release v2p4p12.  Keep in mind that using --chrom with process substitution is inefficient for the same reason that it is with stdin.  --chrom works efficiently with Starch archives and with regular BED files.
I'm going to close this out now.  I will eventually add checks for the case where someone pipes input as stdin but does not tell the program with '-'.
